lib/ASTGen/Sources/ASTGen/Attrs.swift

```diff
@@ -0,0 +1,1132 @@
+      return nil
+    }
+    guard
+      let name = self.generateStringLiteralTextIfNotInterpolated(expr: arg)
+    else {
+      // TODO: Diagnose.
+      return nil
+    }
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      range: self.generateSourceRange(node),
+      name: name
+    )
+  }
+
+  func generateAvailableAttr(attribute node: AttributeSyntax) -> [BridgedAvailableAttr] {
+    guard
+      // `@_OriginallyDefinedIn` has special argument list syntax.
+      let args = node.arguments?.as(AvailabilityArgumentListSyntax.self)
+    else {
+      // TODO: Diagnose.
+      return []
+    }
+
+    fatalError("unimplemented")
+  }
+
+  func generateDynamicReplacementAttr(attribute node: AttributeSyntax) -> BridgedDynamicReplacementAttr? {
+    guard
+      // `@_dynamicReplacement` has special argument list syntax
+      let arg = node.arguments?.as(DynamicReplacementAttributeArgumentsSyntax.self)
+    else {
+      // TODO: Diagnose
+      return nil
+    }
+
+    let replacedFunction = self.generateDeclNameRef(declReferenceExpr: arg.declName)
+
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      attrNameLoc: self.generateSourceLoc(node.attributeName),
+      lParenLoc: self.generateSourceLoc(node.leftParen),
+      replacedFunction: replacedFunction.name,
+      rParenLoc: self.generateSourceLoc(node.rightParen)
+    )
+  }
+
+  func generateEffectsAttr(attribute node: AttributeSyntax) -> BridgedEffectsAttr? {
+    guard
+      let arguments = node.arguments?.as(EffectsAttributeArgumentListSyntax.self),
+      arguments.count >= 1
+    else {
+      // TODO: Diagnose?
+      return nil
+    }
+    let effectKind: BridgedEffectsKind
+    switch arguments.first!.rawText {
+    case "readonly":
+      effectKind = .readOnly
+    case "readnone":
+      effectKind = .readNone
+    case "releasenone":
+      effectKind = .releaseNone
+    case "readwrite":
+      effectKind = .readWrite
+    default:
+      effectKind = .custom
+    }
+
+    if effectKind != .custom {
+      return .createParsed(
+        self.ctx,
+        atLoc: self.generateSourceLoc(node.atSign),
+        range: self.generateSourceRange(node),
+        effectKind: effectKind
+      )
+    } else {
+      return .createParsed(
+        self.ctx,
+        atLoc: self.generateSourceLoc(node.atSign),
+        range: self.generateSourceRange(node),
+        customString: self.extractRawText(arguments).bridged,
+        customStringLoc: self.generateSourceLoc(arguments)
+      )
+    }
+  }
+
+  func generateExclusivityAttr(attribute node: AttributeSyntax) -> BridgedExclusivityAttr? {
+    let mode: BridgedExclusivityAttrMode? = self.generateSingleAttrOption(
+      attribute: node,
+      {
+        switch $0.rawText {
+        case "checked": return .checked
+        case "unchecked": return .unchecked
+        default: return nil
+        }
+      }
+    )
```

Comment:
Would this read nicer with a trailing closure? Same for the other `generateSingleAttrOption` calls.